File: csharp/Hard.cs
using System;
using System.Net;
using System.Web;
using System.Collections.Concurrent;
using System.Threading;

public class Hard
{
    // Insecure Direct Object Reference (IDOR)
    public string GetInvoice(HttpRequest req)
    {
        string invoiceId = req.QueryString["id"];
        // No check if user is authorized to access invoiceId
        return $"Invoice details for {invoiceId}";
    }

    // SSRF vulnerability
    public string FetchUrl(HttpRequest req)
    {
        string url = req.QueryString["url"];
        using (WebClient client = new WebClient())
        {
            return client.DownloadString(url); // No validation on URL
        }
    }

    // Race condition in order processing
    private static ConcurrentDictionary<string, int> inventory = new ConcurrentDictionary<string, int>()
    {
        ["item1"] = 10
    };

    public void BuyItem(HttpRequest req)
    {
        string item = req.QueryString["item"];
        int qty = int.Parse(req.QueryString["qty"]);

        if (inventory.ContainsKey(item) && inventory[item] >= qty)
        {
            // Not atomic update, race condition possible
            inventory[item] -= qty;
        }
    }
}

File: java/hard.java
import java.util.concurrent.*;
import javax.servlet.http.*;

public class Hard {

    public String getInvoice(HttpServletRequest req) {
        String invoiceId = req.getParameter("id");

        return "Invoice contents for " + invoiceId;
    }


    public String fetch(HttpServletRequest req) throws Exception {
        String url = req.getParameter("url");
        java.net.URL u = new java.net.URL(url);
        BufferedReader in = new BufferedReader(new InputStreamReader(u.openStream()));
        return in.readLine();
    }


    private final ConcurrentHashMap<String, Integer> inventory = new ConcurrentHashMap<>();
    public void buyItem(HttpServletRequest req) {
        String item = req.getParameter("item");
        int requested = Integer.parseInt(req.getParameter("qty"));

        if (inventory.getOrDefault(item, 0) >= requested) {
            inventory.put(item, inventory.get(item) - requested);
        }
    }
}
